### PR TITLE
Pre-emptive geofence triggering

### PR DESCRIPTION
**Pull request by jkflying**
**Describe problem solved by this pull request**
Currently, a geofence breach is triggered only after the vehicle is already in violation. This can lead to legal issues if the vehicle ends up outside of the expected area, safety issues if an obstacle is hit that the geofence was meant to avoid, and the vehicle can also be 'stuck' if the breach is continually triggered, overriding attempted user actions.

**Describe your solution**
Predict the motion of the vehicle and trigger the breach if it is determined that the current trajectory would result in a breach happening before the vehicle can make evasive maneuvers, if it delayed any longer. The vehicle is then re-routed to a safe hold position (stop for multirotor, loiter for fixed wing).

**Describe possible alternatives**
There was also the idea of 'shrinking' the allowed flight area, however this would not be robust to different vehicle configurations or tuning.

**Test data / coverage**
Extensive testing both in SITL and on real vehicles (fixed wing, VTOL and multirotor) inside of Auterion. I haven't tested it yet since upstreaming, and additional testing from anyone else would be appreciated.

**Additional context**
Screenshot_20201216_141033

### CODE REVIEW

File: src/modules/navigator/GeofenceBreachAvoidance/fake_geofence.hpp [hunk 1]
```diff
@@ -0,0 +1,183 @@
+
```
Comment (dagar) [inline src/modules/navigator/GeofenceBreachAvoidance/fake_geofence.hpp:1]: Copyright header

File: src/modules/navigator/GeofenceBreachAvoidance/geofence_breach_avoidance.h [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+	} flags;
+	uint8_t value;
+};
+
+class GeofenceBreachAvoidance : public ModuleParams
+{
+public:
+	GeofenceBreachAvoidance();
+
+	~GeofenceBreachAvoidance();
```
Comment (dagar) [inline src/modules/navigator/GeofenceBreachAvoidance/geofence_breach_avoidance.h:56]:
```suggestion
	~GeofenceBreachAvoidance() = default;
```

File: src/modules/navigator/navigator_main.cpp [hunk 3]
```diff
@@ -482,36 +484,112 @@ Navigator::run()
...
-						      home_position_valid());
+			const position_controller_status_s &pos_ctrl_status = _position_controller_status_sub.get();
+
+			matrix::Vector2<double> fence_violation_test_point;
+			geofence_violation_type_u gf_violation_type{};
+			float test_point_bearing;
+			float test_point_distance;
+			float vertical_test_point_distance;
+
+			if (_vstatus.vehicle_type == vehicle_status_s::VEHICLE_TYPE_ROTARY_WING) {
```
Comment (dagar) [inline src/modules/navigator/navigator_main.cpp:495]: Also require `_local_pos.v_xy_valid` here?

File: src/modules/navigator/navigator_main.cpp [hunk 4]
```diff
@@ -482,36 +484,112 @@ Navigator::run()
 		    (_geofence.getGeofenceAction() != geofence_result_s::GF_ACTION_NONE) &&
 		    (hrt_elapsed_time(&last_geofence_check) > GEOFENCE_CHECK_INTERVAL)) {
 
-			bool inside = _geofence.check(_global_pos, _gps_pos, _home_pos,
-						      home_position_valid());
+			const position_controller_status_s &pos_ctrl_status = _position_controller_status_sub.get();
```
Comment (bkueng) [inline src/modules/navigator/navigator_main.cpp:487]: Can you move this block into a separate method?

File: src/modules/navigator/navigator_main.cpp [hunk 5]
```diff
@@ -210,6 +210,8 @@ Navigator::run()
 
 			// update parameters from storage
 			params_update();
+
+			_gf_breach_avoidance.updateParameters();
```
Comment (bkueng) [inline src/modules/navigator/navigator_main.cpp:214]: Instead, pass `this` to `GeofenceBreachAvoidance`: `GeofenceBreachAvoidance(ModuleParams* parent) : ModuleParams(parent) {}`
```suggestion
```

File: src/modules/navigator/GeofenceBreachAvoidance/fake_geofence.hpp [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+class FakeGeofence : public Geofence
+{
+public:
+	FakeGeofence() :
+		Geofence(nullptr)
+	{};
+
+	virtual ~FakeGeofence() {};
+
+	virtual bool isInsidePolygonOrCircle(double lat, double lon, float altitude)
```
Comment (bkueng) [inline src/modules/navigator/GeofenceBreachAvoidance/fake_geofence.hpp:16]:
```suggestion
	bool isInsidePolygonOrCircle(double lat, double lon, float altitude) override
```

File: src/modules/navigator/GeofenceBreachAvoidance/fake_geofence.hpp [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+		case GF_BOUNDARY_20M_AHEAD: {
+				return _gf_boundary_is_20m_ahead(lat, lon, altitude);
+			}
+
+		default:
+			return _allPointsOutside(lat, lon, altitude);
+		}
+	}
+
+	enum PROBE_FUNC_ENUM {
```
Comment (bkueng) [inline src/modules/navigator/GeofenceBreachAvoidance/fake_geofence.hpp:40]:
```suggestion
	enum class ProbeFunction {
```

File: src/modules/navigator/GeofenceBreachAvoidance/fake_geofence.hpp [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+	PROBE_FUNC_ENUM _probe_function_behavior = ALL_POINTS_OUTSIDE;
+
+	bool _allPointsOutside(double lat, double lon, float alt)
+	{
+		return false;
+	}
+
+	bool _left_inside_right_outside(double lat, double lon, float alt)
+	{
+		static int flag = true;
```
Comment (bkueng) [inline src/modules/navigator/GeofenceBreachAvoidance/fake_geofence.hpp:61]: Can you avoid the static and use the correct type?

File: src/modules/navigator/GeofenceBreachAvoidance/fake_geofence.hpp [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+			flag = false;
+			return true;
+
+		} else {
+			flag = true;
+			return false;
+		}
+	}
+
+	bool _gf_boundary_is_20m_ahead(double lat, double lon, float alt)
```
Comment (bkueng) [inline src/modules/navigator/GeofenceBreachAvoidance/fake_geofence.hpp:86]: Might be a detail, but this only does what I'd expect from the method name if yaw=0.

File: src/modules/navigator/GeofenceBreachAvoidance/fake_geofence.hpp [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+	DM_KEY_MISSION_STATE,		/* Persistent mission state */
+	DM_KEY_COMPAT,
+	DM_KEY_NUM_KEYS			/* Total number of item types defined */
+} dm_item_t;
+
+typedef enum {
+	DM_INIT_REASON_POWER_ON = 0,	/* Data survives resets */
+	DM_INIT_REASON_IN_FLIGHT,		/* Data survives in-flight resets only */
+	DM_INIT_REASON_VOLATILE			/* Data does not survive reset */
+} dm_reset_reason;
```
Comment (bkueng) [inline src/modules/navigator/GeofenceBreachAvoidance/fake_geofence.hpp:125]: Can you include the header for these?

File: src/modules/navigator/GeofenceBreachAvoidance/geofence_breach_avoidance.cpp [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+	updateParameters();
+}
+
+GeofenceBreachAvoidance::~GeofenceBreachAvoidance()
+{
+
+}
+
+void GeofenceBreachAvoidance::updateParameters()
+{
```
Comment (bkueng) [inline src/modules/navigator/GeofenceBreachAvoidance/geofence_breach_avoidance.cpp:60]:
```suggestion
void GeofenceBreachAvoidance::updateParams() override
{
    ModuleParams::updateParams();
```

File: src/modules/navigator/GeofenceBreachAvoidance/geofence_breach_avoidance.cpp [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+		float current_max = _test_point_distance;
+		Vector2d test_point;
+
+		// logarithmic search for the distance from the drone to the geofence in the given direction
+		while (abs(current_max - current_min) > 0.5f) {
+			test_point = waypointFromBearingAndDistance(_current_pos_lat_lon, _test_point_bearing, current_distance);
+
+			if (!geofence->isInsidePolygonOrCircle(test_point(0), test_point(1), _current_alt_amsl)) {
+				current_max = current_distance;
+				current_distance = current_min + (current_max - current_min) * 0.5f;
```
Comment (bkueng) [inline src/modules/navigator/GeofenceBreachAvoidance/geofence_breach_avoidance.cpp:170]: This is the same in both cases -> you can move it out of the if and simplify:
```suggestion
				current_distance = (current_max + current_min) * 0.5f;
```

File: src/modules/navigator/GeofenceBreachAvoidance/geofence_breach_avoidance.cpp [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+{
+	const float accel_delay_max = math::max(_params.param_mpc_acc_hor, _params.param_mpc_acc_hor_max);
+	VelocitySmoothing predictor(accel_delay_max, _velocity_hor_abs, 0.f);
+	predictor.setMaxVel(_velocity_hor_abs);
+	predictor.setMaxAccel(_params.param_mpc_acc_hor);
+	predictor.setMaxJerk(_params.param_mpc_jerk_auto);
+	predictor.updateDurations(0.f);
+	predictor.updateTraj(predictor.getTotalTime());
+
+	_multirotor_braking_distance =  predictor.getCurrentPosition() + 0.2f * _velocity_hor_abs; // navigator runs at 5Hz
```
Comment (bkueng) [inline src/modules/navigator/GeofenceBreachAvoidance/geofence_breach_avoidance.cpp:231]: Can you use `GEOFENCE_CHECK_INTERVAL`?

File: src/modules/navigator/GeofenceBreachAvoidance/geofence_breach_avoidance.cpp [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+
+	VelocitySmoothing predictor(accel_delay_max, vertical_vel_abs, 0.f);
+	predictor.setMaxVel(vertical_vel_abs);
+	predictor.setMaxAccel(_climbrate > 0 ? _params.param_mpc_acc_down_max : _params.param_mpc_acc_up_max);
+	predictor.setMaxJerk(_params.param_mpc_jerk_auto);
+	predictor.updateDurations(0.f);
+	predictor.updateTraj(predictor.getTotalTime());
+
+	_multirotor_vertical_braking_distance =  predictor.getCurrentPosition() + 0.2f *
+			vertical_vel_abs; // navigator runs at 5Hz
```
Comment (bkueng) [inline src/modules/navigator/GeofenceBreachAvoidance/geofence_breach_avoidance.cpp:250]: Same here

File: src/modules/navigator/navigator_main.cpp (hunk 3 above)
Comment (jkflying) [reply src/modules/navigator/navigator_main.cpp:495]: This already checks for the GPS being valid via `have_geofence_position_data` , do you think it needs more?

File: src/modules/navigator/GeofenceBreachAvoidance/fake_geofence.hpp (hunk 10 above)
Comment (jkflying) [reply src/modules/navigator/GeofenceBreachAvoidance/fake_geofence.hpp:163]: Moved these to a `dataman_mocks.hpp` as well

File: src/modules/navigator/GeofenceBreachAvoidance/geofence_breach_avoidance.cpp (hunk 12 above)
Comment (jkflying) [reply src/modules/navigator/GeofenceBreachAvoidance/geofence_breach_avoidance.cpp:165]: When you're worried about overflow or floats that implementation can actually be more accurate, but in this case I agree it is overkill.

File: src/modules/navigator/GeofenceBreachAvoidance/fake_geofence.hpp (hunk 9 above)
Comment (jkflying) [reply src/modules/navigator/GeofenceBreachAvoidance/fake_geofence.hpp:124]: renamed to `XXX_north`

**Comment by jkflying** [comment]
All comments addressed

**Review by bkueng — DISMISSED** [review]

**Comment by LorenzMeier** [comment]
@jkflying The geofencebreach test is failing in the PR.

File: src/modules/navigator/navigator_main.cpp (hunk 3 above)
Comment (dagar) [reply src/modules/navigator/navigator_main.cpp]: Depending on `GF_SOURCE` (GPS vs global position).

**Comment by jkflying** [comment]
> @jkflying The geofencebreach test is failing in the PR.

Yep it seems `clang` is more strict with the implicit conversions from float -> double than GCC - it even complains about constants, not just variables. I've done a workaround here, and made a fix in Matrix that will prevent this for the future.

**Comment by garfieldG** [comment]
https://github.com/PX4/PX4-Autopilot/issues/16679

File: src/modules/navigator/navigator_main.cpp (hunk 3 above)
Comment (MikeCharikov) [reply src/modules/navigator/navigator_main.cpp]: @dagar 

Looks like GF_SOURCE logic is broken here. 

https://github.com/PX4/PX4-Autopilot/issues/20677

**Pull request merged**